rapids-cmake/test/install_relocatable.cmake

```diff
@@ -0,0 +1,116 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+rapids_test_install_relocatable
+-------------------------------
+
+.. versionadded:: v23.04.00
+
+Install the needed `ctest` infrastructure to allow installed tests to be run
+by `ctest` in parallel with GPU awareness.
+
+  .. code-block:: cmake
+
+    rapids_test_install_relocatable(INSTALL_COMPONENT_SET <component>
+                                    DESTINATION <relative_path>
+                                    [EXCLUDE_FROM_ALL])
+
+Will install all tests created by :cmake:command:`rapids_test_add` that are
+part of the provided ``INSTALL_COMPONENT_SET``.
+
+The :cmake:command:`rapids_test_install_relocatable` presumes that all
+arguments provided to the tests are machine independent ( no absolute paths ).
+
+``INSTALL_COMPONENT_SET``
+  Record which test component infrastructure to be installed
+
+``DESTINATION``
+  Relative path from the `CMAKE_INSTALL_PREFIX` to install the infrastructure.
+  This needs to be the same directory as the test executables
+
+``EXCLUDE_FROM_ALL``
+  State that these install rules shouldn't be part of the default install set, and
+  instead must be explicitly installed.
+
+
+#]=======================================================================]
+function(rapids_test_install_relocatable)
+  list(APPEND CMAKE_MESSAGE_CONTEXT "rapids.test.install_relocatable")
+
+  set(options EXCLUDE_FROM_ALL)
+  set(one_value INSTALL_COMPONENT_SET DESTINATION)
+  set(multi_value)
+  cmake_parse_arguments(_RAPIDS_TEST "${options}" "${one_value}" "${multi_value}" ${ARGN})
+
+  set(to_exclude)
+  if(_RAPIDS_TEST_EXCLUDE_FROM_ALL)
+    set(to_exclude EXCLUDE_FROM_ALL)
+  endif()
+
+  set(component ${_RAPIDS_TEST_INSTALL_COMPONENT_SET})
+  if(NOT TARGET rapids_test_install_${component})
+    message(FATAL_ERROR "No install component set [${component}] can be found")
+  endif()
+
+  get_target_property(targets_to_install rapids_test_install_${component} TARGETS_TO_INSTALL)
+  get_target_property(tests_to_run rapids_test_install_${component} TESTS_TO_RUN)
+
+  set(content
+      [==[
+  set(CTEST_SCRIPT_DIRECTORY ".")
+  set(CTEST_RESOURCE_SPEC_FILE "./resource_spec.json")
+  execute_process(COMMAND ./generate_ctest_json OUTPUT_FILE "${CTEST_RESOURCE_SPEC_FILE}")
```

Comment:
IIUC this code is necessary because while init.cmake would have generated the resource spec for the machine that the code was built on, this snippet is being embedded in the CMake code that is shipped with the built library so that it can be used to run tests via ctest on an installed binary. Is that right?

If that is right, then I supposed it's not that important for the filenames to match exactly in the two cases, but nonetheless if consistency is the goal could we store the filenames like `resource_spec.json` and `generate_ctest_json` in some centralized constants.json file? The current approach is a bit error-prone w.r.t. modifications needing to happen in both places.

Reply:
Updated with file names moved to a common location